Make partitions memory group aware

StephanDollberg opened the pull request:
Previously `topic_memory_per_partition` was more a rough guess whose
value was way too large. This was partly a result of not having a better
idea but also because obviously not all memory can be used by
partitions.

After some analysis via metrics and the memory profiler we now have a
better idea of what the real value is. Hence, we aggressively change the
value down.

At the same time we make the partition_allocator memory limit check
memory group aware. It will no longer compare
`topic_memory_per_partition` against the total memory as if partitions
could use all that memory. Instead it now compares it against the memory
reserved for partitions via the memory groups.

We use some heuristics (which are better explained in the code comment)
to try to guard against cases where we would make partition density
worse.

The new defaults are:

 - topic_memory_per_partition: 200KiB (this already assumes some to be
   merged optimizations in the seastar metrics stack). It's still fairly
   conservative. Probably more like 150KiB.
 - topic_partitions_max_memory_allocation_share: 10% - together with the
   above this effectively gives us twice the partition density compared
   to the old calculation and using 4MiB for TMPP

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

### Features

* Improved partition density - This release increased the amount of partitions Redpanda can handle. At the same time we also change the behaviour and defaults of the related topic properties `topic_memory_per_partition` (new default 200KiB) and `topic_partitions_per_shard` (new default 5000). To calculate the total allowable amount of partition replicas we no longer divide the total memory space by `topic_memory_per_partition` but rather an explicit amount that is reserved for partitions via a new property `topic_partitions_memory_allocation_percent` (default 10%). This effectively results in allowing double the amount of partitions as previously by default. We heuristically try to detect if these settings were previously overriden and apply the previous logic if so.

StephanDollberg committed:
partition_allocator: Move memory check out of check_limits

That function is getting kinda big. Move the memory checking part out in
preparation for some changes to come.

StephanDollberg committed:
partition_allocator: Don't use binding for topic_memory_per_partition

It's not being watched so no need to use a binding. Also bindings don't
allow using any of the advanced property methods (is_overriden etc.).

StephanDollberg committed:
memory_groups: Introduce reserve for partitions

Static partition memory use is currently unaccounted for in memory
groups.

This patch introduces a reserve to account for partition memory usage.

We don't use a share like some of the other groups but rather an upfront
percentage based. This is because we don't want the partitions share to
shrink once we add more groups.

StephanDollberg committed:
partition_allocator: Make topic_memory_per_partition memory group aware

Previously `topic_memory_per_partition` was more a rough guess whose
value was way too large. This was partly a result of not having a better
idea but also because obviously not all memory can be used by
partitions.

After some analysis via metrics and the memory profiler we now have a
better idea of what the real value is. Hence, we aggressively change the
value down.

At the same time we make the partition_allocator memory limit check
memory group aware. It will no longer compare
`topic_memory_per_partition` against the total memory as if partitions
could use all that memory. Instead it now compares it against the memory
reserved for partitions via using the memory groups.

We use some heuristics (which are better explained in the code comment)
to try to guard against cases where we would make partition density
worse.

The new defaults are:

 - topic_memory_per_partition: 200KiB (this already assumes some to be
   merged optimizations in the seastar metrics stack). It's still fairly
   conservative. Probably more like 150KiB.
 - topic_partitions_max_memory_allocation_share: 10% - together with the
   above this effectively gives us twice the partition density compared
   to the old calculation and using 4MiB for TMPP

StephanDollberg commented:
/ci

StephanDollberg commented:
/dt

StephanDollberg commented:
/dt

StephanDollberg commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59401-->
<!--1658fe1a-b4ad-4c6f-96bf-9979cb17800a-->
<!--title-->
## Retry command for [Build#59401](https://buildkite.com/redpanda/redpanda/builds/59401)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/resource_limits_test.py::ResourceLimitsTest.test_cpu_limited
```
<!--/body-->

StephanDollberg commented:
/dt

vbotbuildovich commented:
<!--RETRIED FAILURES-->
<!--dc9a87f8-7d0b-4e3d-9337-e95fe60fc079-->
<!--body-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/59419#0193aab7-1a96-408b-a78f-e1821ad06d33
<!--/body-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59796-->
<!--e7a26071-f7ee-419c-9d04-9a9e8799265d-->
<!--title-->
## Retry command for [Build#59796](https://buildkite.com/redpanda/redpanda/builds/59796)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--4e0a47dc-6b57-4096-98aa-9165ba167eb9-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59796">build#59796</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59796#0193cf27-358d-4f0b-b9e8-38abf528ebd4 | FAIL          | 0/6      |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3       | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59796#0193cf27-358f-495e-b455-338def428864 | FLAKY         | 3/6      |
</details>
<!--/body-->

Deflaimun commented on src/v/config/configuration.cc:
```diff
@@ -335,6 +336,15 @@ configuration::configuration()
+  , topic_partitions_max_memory_allocation_share(
+      *this,
+      "topic_partitions_max_memory_allocation_share",
+      "foo",
```
placeholder? :D

StephanDollberg replied:
Yes thanks fixed.

JakeSCahill commented on src/v/config/configuration.cc:
```diff
@@ -335,6 +336,16 @@ configuration::configuration()
+  , topic_partitions_memory_allocation_percent(
+      *this,
+      "topic_partitions_memory_allocation_percent",
+      "Percentage of total memory that is being reserved for topic partitions.",
```
```suggestion
      "Percentage of total memory to reserve for topic partitions.",
```

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59813-->
<!--3bf50d4a-e21c-4757-80d1-3d981ac08183-->
<!--e46d8e74-634e-45de-aec8-52c4dafbf5ec-->
<!--title-->
## Retry command for [Build#59813](https://buildkite.com/redpanda/redpanda/builds/59813)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/maintenance_test.py::MaintenanceTest.test_maintenance_sticky@{"use_rpk":true}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--d962d8fe-c3eb-4460-b1f1-42800801aa2e-->
<!--TEST RESULTS-->
<!--a620849b-d289-43ea-a161-0e92438ad6a4-->
<!--TEST RESULTS-->
<!--7a538aae-ab15-4be8-b8d7-b76fc976f272-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59813">build#59813</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59813#0193d07b-7144-4975-a69e-f771323494df | FAIL | 0/6 |
| rptest.tests.maintenance_test.MaintenanceTest.test_maintenance_sticky.use_rpk=True | ducktape | https://buildkite.com/redpanda/redpanda/builds/59813#0193d08e-3cfd-4271-a0b1-b8be6c886960 | FAIL | 0/1 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59905">build#59905</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| gtest_raft_rpunit.gtest_raft_rpunit | unit | https://buildkite.com/redpanda/redpanda/builds/59905#0193d968-7f3b-410b-a73d-f99984280f13 | FLAKY | 1/2 |
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/59905#0193d9ae-342d-4f79-a509-5ad743cfc794 | FAIL | 0/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59905#0193d9a9-d3b3-4212-9ce5-0319f44e90fd | FLAKY | 4/6 |
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/59905#0193d9ae-342f-4b4c-89af-501b30c977b7 | FAIL | 0/6 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/59961">build#59961</a></summary>


| test_id                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cloud_retention_test.CloudRetentionTest.test_cloud_retention.max_consume_rate_mb=None.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/59961#0193dfd7-5068-430e-85d0-afde68da028d | FAIL          | 0/6      |
</details>
<!--/body-->

StephanDollberg commented on src/v/resource_mgmt/memory_groups.h:
```diff
@@ -31,6 +31,12 @@ struct compaction_memory_reservation {
 };
 
+struct partitions_memory_reservation {
+    size_t max_limit_pct = 0;
```
No particular reason. We can change it.

StephanDollberg commented on src/v/cluster/scheduling/partition_allocator.cc:
```diff
@@ -81,6 +81,92 @@ allocation_constraints partition_allocator::default_constraints() {
+    // than 10x smaller. We assume it's unlikely someone would have changed the
+    // value to be that much smaller and hence guess that all the values larger
+    // than 2 times the new default are old values. Equally in the new world
+    // it's unlikely anybody would increase the value (at all really).
```
I mean we can always change the default up reallyas the heuristic is only really needed if the value is overriden (at which point we can also reevaluate the heuristic). 

Lets say we had to change it to 400K from the current 200K then today I feel like the heuristic as is would still be useful as I don't think anyone would have put it below 1MB (and the caveat that nothing immediately breaks still applies).

StephanDollberg commented on src/v/cluster/scheduling/topic_memory_per_partition_default.h:
```diff
@@ -0,0 +1,32 @@
+// had overridden the default when it was non-mmemory group aware. Hence, this
+// value should NOT be changed and stay the same even if the (above) default is
+// changed.
+inline constexpr size_t ORIGINAL_MEMORY_GROUP_AWARE_TMPP = 200_KiB;
```
Yes I guess (see the other comment).

We could also remove this bit of logic I am not sure whether it is overdoing it.

StephanDollberg commented on src/v/cluster/scheduling/partition_allocator.h:
```diff
@@ -34,7 +34,6 @@ class partition_allocator {
     partition_allocator(
       ss::sharded<members_table>&,
       ss::sharded<features::feature_table>&,
-      config::binding<std::optional<size_t>> memory_per_partition,
```
Yes, the problem with bindings is that you can't access the properties of a config like (`is_overriden` etc).

StephanDollberg commented on src/v/cluster/scheduling/topic_memory_per_partition_default.h:
```diff
@@ -0,0 +1,32 @@
+/*
```
Sure. Did you put this comment into this file intentionally?

Bit confused as you mention "Z shares and W total shares" but the partition side doesn't use any shares so wondering whether you think we should print only the partition relevant info (maybe in the partition allocator?) or also all the other memory group info (including shares for other subsystems).

StephanDollberg replied:
> Partitions have 10 shares in memory groups that's what I was referring to. So my comment WXYZ thing was really about memory groups: we should print all the "variable info" from memory groups so we know exactly how the memory was split.

Yeah ok I was just confused because the partitions one doesn't use "shares" but a percentage reservation. 

In any case I have pushed a commit which prints all the final memory group allocations on startup.

Looks like:

```
... [shard 0:main] main - Per shard memory group allocations: total memory: 4.000GiB, total memory minus pre-share reservations: 3.475GiB, chunk cache: 627.953MiB, kafka: 1.226GiB, rpc: 837.271MiB, recovery: 418.635MiB, tiered storage: 418.635MiB, data transforms: 0.000bytes, compaction: 128.000MiB, datalake: 0.000bytes, partitions: 409.600MiB
```

> We should also have good logging about partition allocations decisions (and failure reasons), but I think we already have that?

Yes the updated failure warning should have all the details now.

StephanDollberg commented on src/v/cluster/scheduling/partition_allocator.cc:
```diff
@@ -81,6 +81,92 @@ allocation_constraints partition_allocator::default_constraints() {
+bool guess_is_memory_group_aware_memory_limit() {
+    // Here we are trying to guess whether `topic_memory_per_partition` is
+    // memory group aware. Originally the property was more like a guess of how
+    // much memory each partition would statically use. It was first set to 10
```
Ah now that you say it that rings a bell. I misremembered.

StephanDollberg commented on src/v/cluster/scheduling/partition_allocator.h:
```diff
@@ -149,6 +148,12 @@ class partition_allocator {
 
+    // sub-routine of the above, checks available memory
+    std::error_code check_memory_limits(
+      const uint64_t new_partitions_replicas_requested,
```
removed

StephanDollberg commented on src/v/cluster/scheduling/partition_allocator.cc:
```diff
@@ -81,6 +81,92 @@ allocation_constraints partition_allocator::default_constraints() {
+          clusterlog.warn,
+          "Refusing to create {} new partitions as total partition count "
+          "{} "
+          "would exceed memory limit {}",
```
Added a few more. Let me know if you think anything is missing.

StephanDollberg commented on src/v/cluster/scheduling/partition_allocator.cc:
```diff
@@ -81,6 +81,92 @@ allocation_constraints partition_allocator::default_constraints() {
 
+bool guess_is_memory_group_aware_memory_limit() {
+    // Here we are trying to guess whether `topic_memory_per_partition` is
+    // memory group aware. Originally the property was more like a guess of how
```
changed

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59905-->
<!--8d5c07de-0100-4c54-af59-429ae99a1fba-->
<!--3b3c29b9-eef2-4c73-9665-292f2e8e3612-->
<!--title-->
## Retry command for [Build#59905](https://buildkite.com/redpanda/redpanda/builds/59905)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```

/ci-repeat 1
tests/rptest/tests/datalake/partition_movement_test.py::PartitionMovementTest.test_cross_core_movements@{"cloud_storage_type":1}
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

StephanDollberg committed:
memory_groups: Add startup log of all memory group allocations

Adds a log to startup printing all the final memory group allocations to
the different sub systems.

Logs like:

```
... [shard 0:main] main - Per shard memory group allocations: total memory: 4.000GiB, total memory minus pre-share reservations: 3.475GiB, chunk cache: 627.953MiB, kafka: 1.226GiB, rpc: 837.271MiB, recovery: 418.635MiB, tiered storage: 418.635MiB, data transforms: 0.000bytes, compaction: 128.000MiB, datalake: 0.000bytes, partitions: 409.600MiB
```

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/59961-->
<!--717ac6d4-bc40-42da-8a87-82ec90e222d7-->
<!--title-->
## Retry command for [Build#59961](https://buildkite.com/redpanda/redpanda/builds/59961)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cloud_retention_test.py::CloudRetentionTest.test_cloud_retention@{"cloud_storage_type":2,"max_consume_rate_mb":null}
```
<!--/body-->

dotnwat commented on src/v/resource_mgmt/memory_groups.cc:
```diff
@@ -140,6 +141,26 @@ size_t system_memory_groups::total_memory() const {
 
+void system_memory_groups::log_memory_group_allocations(seastar::logger& log) {
+    log.info(
+      "Per shard memory group allocations: total memory: {}, "
+      "total memory minus pre-share reservations: {}, chunk cache: {}, kafka: "
+      "{}, rpc: {}, recovery: {}, "
+      "tiered storage: {}, data transforms: {}, compaction: {}, datalake: {}, "
+      "partitions: {}",
```
cool
